File: comic-share/frontend/src/services/llm.ts
/**
 * LLM 服务 — SSE 流式调用。
 */

import { logger } from '@/utils/logger'

const API_BASE = import.meta.env.VITE_API_BASE_URL || ''

export interface LLMGenerateParams {
  prompt: string
  system_prompt?: string
  provider?: string
  model?: string
  api_key?: string
  temperature?: number
  max_tokens?: number
}

export interface LLMModel {
  id: string
  name: string
}

/**
 * 获取模型列表
 */
export async function fetchModels(provider?: string): Promise<LLMModel[]> {
  const resp = await fetch(`${API_BASE}/api/v1/llm/models`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ provider }),
  })
  if (!resp.ok) throw new Error(`获取模型失败: ${resp.status}`)
  const data = await resp.json()
  return data.models || []
}

/**
 * 流式生成文本，回调接收每个 chunk
 */
export async function generateStream(
  params: LLMGenerateParams,
  onChunk: (text: string) => void,
  onDone?: (fullText: string) => void,
  onError?: (error: string) => void,
): Promise<void> {
  const resp = await fetch(`${API_BASE}/api/v1/llm/generate`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })

  if (!resp.ok) {
    const errText = await resp.text()
    onError?.(errText)
    return
  }

  const reader = resp.body?.getReader()
  if (!reader) {
    onError?.('无法读取响应流')
    return
  }

  const decoder = new TextDecoder()
  let buffer = ''

  try {
    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const lines = buffer.split('\n')
      buffer = lines.pop() || ''

      for (const line of lines) {
        if (line.startsWith('event: ')) {
          continue
        }
        if (line.startsWith('data: ')) {
          const jsonStr = line.slice(6)
          try {
            const data = JSON.parse(jsonStr)
            if (data.content !== undefined) {
              onChunk(data.content)
            } else if (data.full_response !== undefined) {
              onDone?.(data.full_response)
            } else if (data.error !== undefined) {
              onError?.(data.error)
            }
          } catch {
            logger.debug('SSE JSON parse skip:', jsonStr)
          }
        }
      }
    }
  } finally {
    reader.releaseLock()
  }
}

File: comic-share/backend/routers/llm.py
"""LLM 路由 — 模型列表 + 流式生成。"""

import logging

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from sse_starlette.sse import EventSourceResponse

from services.llm_service import generate_stream, list_models

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/v1/llm", tags=["LLM"])


class ModelsRequest(BaseModel):
    provider: str | None = None


class GenerateRequest(BaseModel):
    prompt: str
    system_prompt: str = ""
    provider: str | None = None
    model: str | None = None
    api_key: str | None = None
    temperature: float = 0.7
    max_tokens: int = 4096


@router.post("/models")
async def get_models(req: ModelsRequest):
    """获取指定 Provider 的可用模型列表。"""
    try:
        models = await list_models(req.provider)
        return {"models": models}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error("获取模型列表异常: %s", str(e))
        raise HTTPException(status_code=502, detail="无法连接 LLM 服务")


@router.post("/generate")
async def generate(req: GenerateRequest):
    """流式生成文本，SSE 格式输出。"""
    import json

    async def event_generator():
        full_text = ""
        try:
            async for chunk in generate_stream(
                prompt=req.prompt,
                system_prompt=req.system_prompt,
                provider=req.provider,
                model=req.model,
                api_key=req.api_key,
                temperature=req.temperature,
                max_tokens=req.max_tokens,
            ):
                if chunk:
                    full_text += chunk
                    yield {"event": "chunk", "data": json.dumps({"content": chunk})}
            yield {"event": "done", "data": json.dumps({"full_response": full_text})}
        except RuntimeError as e:
            yield {"event": "error", "data": json.dumps({"error": str(e)})}
        except Exception as e:
            logger.error("LLM 生成异常: %s", str(e))
            yield {"event": "error", "data": json.dumps({"error": "内部错误"})}

    return EventSourceResponse(event_generator())
